packages/gatsby/cache-dir/api-runner-ssr.js

```diff
@@ -6,10 +6,45 @@
 
 const apis = require(`./api-ssr-docs`)
 
-module.exports = (api, args, defaultReturn) => {
+/**
+ * Some apis should only be implemented once. Given a list of plugins, and an
+ * `api` to check, this will return [] for apis that are implemented 1 or 0 times.
+ *
+ * For apis that have been implemented multiple times, return an array of paths
+ * pointing to the file implementing `api`.
+ *
+ * @param {array} pluginList
+ * @param {string} api
+ */
+const duplicatedApis = (pluginList, api) => {
+  let implementsApi = []
+
+  pluginList.forEach(p => {
+    if (p.plugin[api]) implementsApi.push(p.path)
+  })
+
+  if (implementsApi.length < 2) return [] // no dupes
+  return implementsApi // paths to dupes
+}
+
+// Run the specified api in any plugins that have implemented it
+const apiRunner = ({ api, args, defaultReturn, checkDupes = false }) => {
   if (!apis[api]) {
     console.log(`This API doesn't exist`, api)
   }
+
+  if (checkDupes) {
+    const dupes = duplicatedApis(plugins, api)
+    if (dupes.length > 0) {
```

Comment:
It looks like you need to filter this list to APIs where duplicates is a problem e.g. `replaceRenderer`. Most SSR APIs are fine.

Reply:
Also, I think we should suppress warnings for duplicated APIs where there's an implementation in the site's gatsby-* file as those always override other plugins. This is the normal fix — merge plugin's implementations of `replaceRenderer`. As things are now, the warning would keep showing which isn't what we want as then the user might uninstall the plugin to fix things but which would actually probably break things as many plugins with `gatsby-ssr.js` files also have other API implementations e.g. for browser APIs.

Reply:
> It looks like you need to filter this list to APIs where duplicates is a problem

This is handled at the point that apiRunner is called, using the new `checkDupes` flag:
```
// check for duplicate api implementations
apiRunner({api: `replaceRenderer`, args, checkDupes: true})

// don't check for duplicate api implementations (default behaviour)
apiRunner({api: `replaceRenderer`, args})
```

But actually it's just the `replaceRenderer` api that shouldn't be duplicated? For now I can hardcode the name `replaceRenderer` within `apiRunner`. Then we can extract it out later if necessary. 

> we should suppress warnings for duplicated APIs where there's an implementation in the site's gatsby-* file

👍